### DISCUSSION TITLE
Routes Defined in Extensions Fail

### DISCUSSION THREAD

**Discussion by dajinchu** [vikejs/vike#893] [Discussion] [answered]
### Description

Apologies if this is intentional behavior. 

As of `0.4.122`, It seems adding routes/page files as part of `extensions` in `vite.config.ts` fails. 

When running in dev mode, I get `No onRenderHtml() hook found` when visiting the page.

[Using the restack project for reproduction
](https://github.com/brillout/restack/compare/main...dajinchu:restack:extensions-page-repro)
full trace:
```
    at executeOnRenderHtmlHook (/workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.123_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/runtime/renderPage/executeOnRenderHtmlHook.js:32:29)
    at renderPageContext (/workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.123_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/runtime/renderPage/renderPageContext.js:60:90)
    at async renderPageAttempt (/workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.123_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/runtime/renderPage.js:163:36)
    at async renderPage (/workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.123_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/runtime/renderPage.js:41:35)
    at async /workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.123_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/shared/addSsrMiddleware.js:31:27
```

Works fine on `0.4.121`, though on all versions, running `vite build` fails with a different error.

```
node:internal/process/promises:289
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error: [vite-plugin-ssr@0.4.121][Bug] You stumbled upon a bug in vite-plugin-ssr's source code. Go to https://github.com/brillout/vite-plugin-ssr/issues/new and copy-paste this error. (The error's stack trace is usually enough to fix the problem). A maintainer will fix the bug (usually under 24 hours). Don't hesitate to reach out as it makes vite-plugin-ssr more robust.
    at getEntryFromPageConfigData (/workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.121_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/buildConfig.js:134:24)
    at /workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.121_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/buildConfig.js:86:48
    at Array.forEach (<anonymous>)
    at analyzeClientEntries (/workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.121_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/buildConfig.js:77:21)
    at getEntries (/workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.121_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/buildConfig.js:52:69)
    at async configResolved (/workspaces/restack/node_modules/.pnpm/vite-plugin-ssr@0.4.121_vite@4.2.1/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/buildConfig.js:18:21)
    at async Promise.all (index 13)
    at async resolveConfig (file:///workspaces/restack/node_modules/.pnpm/vite@4.2.1_@types+node@18.15.3/node_modules/vite/dist/node/chunks/dep-79892de8.js:63716:5)
    at async build (file:///workspaces/restack/node_modules/.pnpm/vite@4.2.1_@types+node@18.15.3/node_modules/vite/dist/node/chunks/dep-79892de8.js:46088:20)
    at async main (file:///workspaces/restack/restack/dist/build-FWXHCI4U.js:9:3)
```

Its a weird use-case so I understand if this is not meant to be supported :)

**Comment by brillout** [comment]
What's your motivation for using `extensions`?

**Comment by dajinchu** [comment]
I'm building a framework for incremental migration off of server-rendered Rails.

The goal is to be able to write new pages fully in React, but any requests that don't match a VPS-defined route get proxied to the legacy Rails app and the returned HTML is inserted into React (layout/navbar lives in React).

I have pages defined in order to switch onRenderHtml/onRenderClient based on if we're serving proxied pages or not. There is some custom server logic to handle the proxying, but its all working if I don't use `extensions`. So I can make this work w/o it, but (a) I'd like to open source, and (b) keep the framework code separate from the application.


BTW, VPS has been amazing. I thought for a niche use-case like this we'd have to drop down to Vite's native SSR, which would have likely been too much work for us, but so far VPS's hooks have been exactly what we need. In the process of getting my company to sponsor :)

**Reply by brillout** [reply]
That's a great use case. Definitely keep me updated about it (also if you run into any fundamental blockers and/or minor issues).

I'm actually migrating away from `extensions` in favor of [`extends`](https://github.com/brillout/vite-plugin-ssr/blob/d265a460d975455c3660822424cd34469c384d3c/examples/vike-react-simple/pages/%2Bconfig.js#L6). It should allow you to achieve the same.

Note that it's all experimental; it's expected that there may be breaking changes in semver minor updates.

Thanks for starting the process of sponsoring, it makes a *huge* difference.

Let me have a look at the reproduction.

**Reply by dajinchu** [reply]
Trying out `extends` a little bit. Just wondering how it would support defining pages? It seems like for each page the framework wants to define, you would have to create a `+config.ts` file on the application side?

Is that intended?

So like, I can define `/test` in the framework, but the application will still need to define `pages/test/+config.ts` and do `extends` inside there?

**Reply by brillout** [reply]
The plan is to implement https://github.com/brillout/vite-plugin-ssr/issues/341 which means that the framework's `+config.js` wil also be able to define pages.

Btw., once sponsoring is set up and you're a top-tier sponsor, I'll add your use case to vite-plugin-ssr's CI in order to ensure that no version will break your framework. I'll also prioritize the roadmap to support your company's endeavours. Do you have an ETA for the sponsoring process? Do you know what tier your company will select?

**Reply by dajinchu** [reply]
Should be next week. Not sure on tier, but getting into CI would be a major selling point. Will relay that to my CTO next week.

**Reply by brillout** [reply]
👍 looking forward to it.

Also your company is using VPS features that other frameworks don't offer so, even without feature prioritization and CI integration, it makes business sense to sponsor VPS.

**Comment by brillout** [comment]
> When running in dev mode, I get `No onRenderHtml() hook found` when visiting the page.

I can't reproduce at https://github.com/brillout/restack/tree/reprod/examples/basic: while I do get a `404` when I go to `/test`, I don't get `No onRenderHtml() hook found`.

Let's start with the regression. What's your reproduction steps?

**Reply by dajinchu** [reply]
Just reproduced with fresh git clone:
1. Checkout `reprod` branch
2. `pnpm i` and `pnpm run build` in restack folder
3. `pnpm run dev` in examples/basic
4. Visit /test
5. `No onRenderHtml() hook found` in server logs

**Reply by brillout** [reply]
~~Still can't reproduce:~~ Ah, indeed, I could reproduce.

```
git clone git@github.com:brillout/restack
cd restack/
git switch reprod
pnpm install
pnpm run build
cd examples/basic/
pnpm run dev
```

Same as single line (copy-paste me):
```
git clone git@github.com:brillout/restack && cd restack/ && git switch reprod && pnpm install && pnpm run build && cd examples/basic/ && pnpm run dev
```

I'll have a look at it today.

**Reply by brillout** [reply, marked as answer]
@dajinchu Just released `0.4.125`: it should now work.

**Reply by dajinchu** [reply]
Works great, thanks!
